api/endpoints.go

```diff
@@ -0,0 +1,302 @@
+	StructFn StructFn `json:"-"`
+
+	// RecordFn is for returning a database record. It will be properly locked
+	// and marshalled including metadata.
+	RecordFn RecordFn `json:"-"`
+
+	// HandlerFn is the raw http handler.
+	HandlerFn http.HandlerFunc `json:"-"`
+}
+
+type (
+	// ActionFn is for simple actions with a return message for the user.
+	ActionFn func(ar *Request) (msg string, err error)
+
+	// DataFn is for returning raw data that the caller for further processing.
+	DataFn func(ar *Request) (data []byte, err error)
+
+	// StructFn is for returning any kind of struct.
+	StructFn func(ar *Request) (i interface{}, err error)
+
+	// RecordFn is for returning a database record. It will be properly locked
+	// and marshalled including metadata.
+	RecordFn func(ar *Request) (r record.Record, err error)
+)
+
+// MIME Types
+const (
+	MimeTypeJSON string = "application/json"
+	MimeTypeText string = "text/plain"
+
+	apiV1Path = "/api/v1/"
+)
+
+func init() {
+	RegisterHandler(apiV1Path+"{endpointPath:.+}", &endpointHandler{})
+}
+
+var (
+	endpoints     = make(map[string]*Endpoint)
+	endpointsLock sync.RWMutex
+
+	// ErrInvalidEndpoint is returned when an invalid endpoint is registered.
+	ErrInvalidEndpoint = errors.New("endpoint is invalid")
+
+	// ErrAlreadyRegistered is returned when there already is an endpoint with
+	// the same path registered.
+	ErrAlreadyRegistered = errors.New("an endpoint for this path is already registered")
+)
+
+func getAPIContext(r *http.Request) (apiEndpoint *Endpoint, apiRequest *Request) {
+	// Get request context and check if we already have an action cached.
+	apiRequest = GetAPIRequest(r)
+	if apiRequest == nil {
+		return nil, nil
+	}
+	var ok bool
+	apiEndpoint, ok = apiRequest.HandlerCache.(*Endpoint)
+	if ok {
+		return apiEndpoint, apiRequest
+	}
+
+	// If not, get the action from the registry.
+	endpointPath, ok := apiRequest.URLVars["endpointPath"]
+	if !ok {
+		return nil, apiRequest
+	}
+
+	endpointsLock.RLock()
+	defer endpointsLock.RUnlock()
+
+	apiEndpoint, ok = endpoints[endpointPath]
+	if ok {
+		// Cache for next operation.
+		apiRequest.HandlerCache = apiEndpoint
+	}
+	return apiEndpoint, apiRequest
+}
+
+// RegisterEndpoint registers a new endpoint. An error will be returned if it
+// does not pass the sanity checks.
+func RegisterEndpoint(e Endpoint) error {
+	if err := e.check(); err != nil {
+		return fmt.Errorf("%w: %s", ErrInvalidEndpoint, err)
+	}
+
+	endpointsLock.Lock()
+	defer endpointsLock.Unlock()
+
+	_, ok := endpoints[e.Path]
+	if ok {
+		return ErrAlreadyRegistered
+	}
+
+	endpoints[e.Path] = &e
+	return nil
+}
+
+func (e *Endpoint) check() error {
+	// Check path.
+	if e.Path == "" {
+		return errors.New("path is missing")
+	}
+
+	// Check permissions.
+	if e.Read < Require || e.Read > PermitSelf {
+		return errors.New("invalid read permission")
+	}
+	if e.Write < Require || e.Write > PermitSelf {
+		return errors.New("invalid write permission")
+	}
```

Comment:
We continue to repeat the "if condition" for checking the validity. Maybe use a helper like `validatePermission() error`